---
title: 'AD Proteomics'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: '../`r unlist(strsplit(getwd(), "/"))[4]`.bib'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "results", "AD-proteomics.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [AD-CBD-exosome-proteins](https://github.com/eturkes/AD-CBD-exosome-proteins). Please email me for access.*

The data here is derived from @`r unlist(strsplit(getwd(), "/"))[4]` and will be referenced using the name `r unlist(strsplit(getwd(), "/"))[4]`.

```{r, boilerplate}
# Copyright 2019-2020 Emir Turkes
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# These should be checked per document.
# -------------------------------------

packages <- c("proteus", "Amelia", "DT")
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path(getwd(), "utils.R"))

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------

data_name <- unlist(strsplit(getwd(), "/"))[4] # Name of dataset.
assets_dir <- file.path(getwd(), "..", "assets") # Backed up caches and data.
tmp_dir <- file.path(getwd(), "..", "tmp")
results_dir <- file.path(getwd(), "..", "results")

if (!dir.exists(tmp_dir)) {
  dir.create(tmp_dir, recursive = TRUE)
}

# Unique cache directory for each analysis number.
if (!dir.exists(file.path(assets_dir, "cache"))) {
  dir.create(file.path(assets_dir, "cache"), recursive = TRUE)
}
```

# All Data

## Prep

```{r, all-analysis}
data <- read.csv(
  file.path(
    assets_dir, "Label_Free_Quantitative_Proteomics_Analyses_Results", "proteinGroups.txt"
  ),
  sep = "\t"
)

# Reverse order of sample columns to make DE analysis more intuitive later.
col_start <- grep("AD_5333", x = colnames(data))
for (i in 1:length(col_start)) {
  section1_end <- col_start[i] - 1
  col_end <- col_start[i] + 8
  section2_start <- col_end + 1
  data <- data.frame(
    data[ , 1:section1_end], rev(data[ , col_start[i]:col_end]), data[ , section2_start:ncol(data)]
  )
}

meta <- data.frame(
  "sample" = gsub("Peptides.", "", colnames(data)[13:21]),
  "conditions" = c(rep("Contr", 4), rep("AD", 5))
)

data <- readProteinGroups(
  file.path(
    assets_dir, "Label_Free_Quantitative_Proteomics_Analyses_Results", "proteinGroups.txt"
  ),
  meta
)

summary(data)
```

## QC {.tabset}

```{r, all-qc}
plotCount(data)
plotDetectionSimilarity(data, bin.size = 0.02)
missmap(as.data.frame(data$tab))
```

### Pre-normalization

```{r, all-pre-norm}
plotSampleDistributions(data, fill = "condition")
proteus::plotPCA(data)
```

### Normalized {.active}

```{r, all-norm}
data <- normalizeData(data)
plotSampleDistributions(data, fill = "condition")
proteus::plotPCA(data)
```

## Modeling

```{r, all-clust}
plotMV(data, with.loess = TRUE)
plotDistanceMatrix(data)
plotClustering(data)
```

## Annotation

```{r, all-anno}
rds <- file.path(assets_dir, "cache", "all_anno.rds")
if (file.exists(rds)) {
  anno_id <- readRDS(rds)
} else {
  ids <- lapply(as.character(data$proteins), function(prot) {
    protein <- unlist(strsplit(prot, ";", fixed = TRUE))[1]
  })
  ids <- data.frame(protein = do.call(rbind, ids))
  anno <- fetchFromUniProt(ids$protein)
  anno_id <- merge(ids, anno, by.x = "protein", by.y = "id")
  anno_id <- anno_id[-which(duplicated(anno_id$genes)), ]
  saveRDS(anno_id, rds)
}
data$proteins <- data$proteins[data$proteins %in% anno_id$protein]
anno_id <- anno_id[anno_id$protein %in% data$proteins, ]
anno_id <- anno_id[order(match(anno_id$protein, data$proteins)), ]
data <- annotateProteins(data, anno_id)
```

## Differential Expression

```{r, all-de}
results <- limmaDE(data)
results <- results[order(results$adj.P.Val), ]
rownames(results) <- NULL
datatable_custom(results[c("protein", "genes", "logFC", "adj.P.Val")])
```

# Outlier Removal

AD_5521 appears to be an outlier, so we remove it and see if it improves statistical power.

## Prep

```{r, outlier-analysis}
data <- read.csv(
  file.path(
    assets_dir, "Label_Free_Quantitative_Proteomics_Analyses_Results", "proteinGroups.txt"
  ),
  sep = "\t"
)

# Reverse order of sample columns to make DE analysis more intuitive later.
col_start <- grep("AD_5333", x = colnames(data))
for (i in 1:length(col_start)) {
  section1_end <- col_start[i] - 1
  col_end <- col_start[i] + 8
  section2_start <- col_end + 1
  data <- data.frame(
    data[ , 1:section1_end], rev(data[ , col_start[i]:col_end]), data[ , section2_start:ncol(data)]
  )
}

data[ , grep("AD_5521", x = colnames(data))] <- NULL

meta <- data.frame(
  "sample" = gsub("Peptides.", "", colnames(data)[13:20]),
  "conditions" = c(rep("Contr", 4), rep("AD", 4))
)

data <- readProteinGroups(
  file.path(
    assets_dir, "Label_Free_Quantitative_Proteomics_Analyses_Results", "proteinGroups.txt"
  ),
  meta
)

summary(data)
```

## QC {.tabset}

```{r, outlier-qc}
plotCount(data)
plotDetectionSimilarity(data, bin.size = 0.02)
missmap(as.data.frame(data$tab))
```

### Pre-normalization

```{r, outlier-pre-norm}
plotSampleDistributions(data, fill = "condition")
proteus::plotPCA(data)
```

### Normalized {.active}

```{r, outlier-norm}
data <- normalizeData(data)
plotSampleDistributions(data, fill = "condition")
proteus::plotPCA(data)
```

## Modeling

```{r, outlier-clust}
plotMV(data, with.loess = TRUE)
plotDistanceMatrix(data)
plotClustering(data)
```

## Annotation

```{r, outlier-anno}
rds <- file.path(assets_dir, "cache", "outlier_anno.rds")
if (file.exists(rds)) {
  anno_id <- readRDS(rds)
} else {
  ids <- lapply(as.character(data$proteins), function(prot) {
    protein <- unlist(strsplit(prot, ";", fixed = TRUE))[1]
  })
  ids <- data.frame(protein = do.call(rbind, ids))
  anno <- fetchFromUniProt(ids$protein)
  anno_id <- merge(ids, anno, by.x = "protein", by.y = "id")
  anno_id <- anno_id[-which(duplicated(anno_id$genes)), ]
  saveRDS(anno_id, rds)
}
data$proteins <- data$proteins[data$proteins %in% anno_id$protein]
anno_id <- anno_id[anno_id$protein %in% data$proteins, ]
anno_id <- anno_id[order(match(anno_id$protein, data$proteins)), ]
data <- annotateProteins(data, anno_id)
```

## Differential Expression

```{r, outlier-de}
results <- limmaDE(data)
results <- results[order(results$adj.P.Val), ]
rownames(results) <- NULL
datatable_custom(results[c("protein", "genes", "logFC", "adj.P.Val")])
```

# References

This is the concluding section of the document. Here we write relevant results to disk, output the `sessionInfo`, and create a bibliography for works cited.

```{r, references}
sessionInfo()
```
