---
title: 'AD Proteomics'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: '../`r unlist(strsplit(getwd(), "/"))[4]`.bib'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "results", "AD-proteomics.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [AD-CBD-exosome-proteins](https://github.com/eturkes/AD-CBD-exosome-proteins). Please email me for access.*

The data here is derived from @`r unlist(strsplit(getwd(), "/"))[4]` and will be referenced using the name `r unlist(strsplit(getwd(), "/"))[4]`.

```{r, boilerplate}
# Copyright 2019-2020 Emir Turkes
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# These should be checked per document.
# -------------------------------------

packages <- c("dplyr", "DEP", "ggplot2", "DT")
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path(getwd(), "utils.R"))

options(stringsAsFactors = FALSE)

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------

data_name <- unlist(strsplit(getwd(), "/"))[4] # Name of dataset.
assets_dir <- file.path(getwd(), "..", "assets") # Backed up caches and data.
tmp_dir <- file.path(getwd(), "..", "tmp")
results_dir <- file.path(getwd(), "..", "results")

if (!dir.exists(tmp_dir)) {
  dir.create(tmp_dir, recursive = TRUE)
}

# Unique cache directory for each analysis number.
if (!dir.exists(file.path(assets_dir, "cache"))) {
  dir.create(file.path(assets_dir, "cache"), recursive = TRUE)
}
```

# All Data

```{r, all-prep}
data <- read.csv(
  file.path(
    assets_dir, "Label_Free_Quantitative_Proteomics_Analyses_Results", "proteinGroups.txt"
  ),
  sep = "\t"
)

# Reverse order of sample columns to make DE analysis more intuitive later.
col_start <- grep("AD_5333", x = colnames(data))
for (i in 1:length(col_start)) {
  section1_end <- col_start[i] - 1
  col_end <- col_start[i] + 8
  section2_start <- col_end + 1
  data <- data.frame(
    data[ , 1:section1_end], rev(data[ , col_start[i]:col_end]), data[ , section2_start:ncol(data)]
  )
}

data <- filter(data, Reverse != "+", Potential.contaminant != "+")
data <- make_unique(data, "Gene.names", "Protein.IDs", delim = ";")
LFQ_columns <- grep("Intensity.", colnames(data))
label = gsub("Peptides.", "", colnames(data)[13:21])
experimental_design <- data.frame(
  label = label, condition = c(rep("Contr", 4), rep("AD", 5)), replicate = label
)

data <- make_se(data, LFQ_columns, experimental_design)
data
```

## QC

```{r, all-qc}
plot_numbers(data)
plot_frequency(data)
plot_missval(data)
data <- filter_proteins(data, "condition", thr = 1)
plot_numbers(data)
plot_frequency(data)
plot_missval(data)

meanSdPlot(data)
data_norm <- normalize_vsn(data)
plot_normalization(data, data_norm)
data <- data_norm
rm(data_norm)
meanSdPlot(data)

plot_detect(data)
set.seed(1)
data_imp <- impute(data, fun = "QRILC")
plot_imputation(data, data_imp)
data <- data_imp
rm(data_imp)
```

## Differential Expression

```{r, all-de}
data <- add_rejections(test_diff(data, type = "control", control = "Contr"), lfc = 0)
data_results <- get_results(data)
nsig <- data_results %>% filter(significant) %>% nrow()

plot_pca(data, n = round(dim(data)[1] * 0.1)) +
  scale_shape_manual(values = seq(0, length(label)))
plot_heatmap(data, type = "centered", kmeans = TRUE, k = round((nsig -1) / 2), row_font_size = 9)
plot_cor(data, pal = "RdBu", pal_rev = TRUE)
plot_volcano(data, contrast = "AD_vs_Contr", adjusted = TRUE, label_size = 4)

plot_single(data, proteins = c("MAPT", "APOE"), type = "centered") +
  xlab(element_blank())

datatable_custom(data_results)
```

# Outlier Removal

AD Replicate 1 (`AD_5521`) appears to be an outlier, so we remove it and see if it improves statistical power.

```{r, outlier-prep}
data <- read.csv(
  file.path(
    assets_dir, "Label_Free_Quantitative_Proteomics_Analyses_Results", "proteinGroups.txt"
  ),
  sep = "\t"
)

# Reverse order of sample columns to make DE analysis more intuitive later.
col_start <- grep("AD_5333", x = colnames(data))
for (i in 1:length(col_start)) {
  section1_end <- col_start[i] - 1
  col_end <- col_start[i] + 8
  section2_start <- col_end + 1
  data <- data.frame(
    data[ , 1:section1_end], rev(data[ , col_start[i]:col_end]), data[ , section2_start:ncol(data)]
  )
}

data[ , grep("AD_5521", x = colnames(data))] <- NULL

data <- filter(data, Reverse != "+", Potential.contaminant != "+")
data <- make_unique(data, "Gene.names", "Protein.IDs", delim = ";")
LFQ_columns <- grep("Intensity.", colnames(data))
label = gsub("Peptides.", "", colnames(data)[13:20])
experimental_design <- data.frame(
  label = label, condition = c(rep("Contr", 4), rep("AD", 4)), replicate = label
)

data <- make_se(data, LFQ_columns, experimental_design)
data
```

## QC

```{r, outlier-qc}
plot_numbers(data)
plot_frequency(data)
plot_missval(data)
data <- filter_proteins(data, "condition", thr = 1)
plot_numbers(data)
plot_frequency(data)
plot_missval(data)

meanSdPlot(data)
data_norm <- normalize_vsn(data)
plot_normalization(data, data_norm)
data <- data_norm
rm(data_norm)
meanSdPlot(data)

plot_detect(data)
set.seed(1)
data_imp <- impute(data, fun = "QRILC")
plot_imputation(data, data_imp)
data <- data_imp
rm(data_imp)
```

## Differential Expression

```{r, outlier-de}
data <- add_rejections(test_diff(data, type = "control", control = "Contr"), lfc = 0)
data_results <- get_results(data)
nsig <- data_results %>% filter(significant) %>% nrow()

plot_pca(data, n = round(dim(data)[1] * 0.1)) +
  scale_shape_manual(values = seq(0, length(label)))
plot_heatmap(data, type = "centered", kmeans = TRUE, k = round((nsig -1) / 2), row_font_size = 9)
plot_cor(data, pal = "RdBu", pal_rev = TRUE)
plot_volcano(data, contrast = "AD_vs_Contr", adjusted = TRUE, label_size = 4)

plot_single(data, proteins = c("MAPT", "APOE"), type = "centered") +
  xlab(element_blank())

datatable_custom(data_results)
```

# More Outliers

We also try removing `AD_5475` and `Contr_5382`, in addition to `AD_5521` from before.

```{r, outlier2-prep}
data <- read.csv(
  file.path(
    assets_dir, "Label_Free_Quantitative_Proteomics_Analyses_Results", "proteinGroups.txt"
  ),
  sep = "\t"
)

# Reverse order of sample columns to make DE analysis more intuitive later.
col_start <- grep("AD_5333", x = colnames(data))
for (i in 1:length(col_start)) {
  section1_end <- col_start[i] - 1
  col_end <- col_start[i] + 8
  section2_start <- col_end + 1
  data <- data.frame(
    data[ , 1:section1_end], rev(data[ , col_start[i]:col_end]), data[ , section2_start:ncol(data)]
  )
}

data[ , grep("AD_5521", x = colnames(data))] <- NULL
data[ , grep("AD_5475", x = colnames(data))] <- NULL
data[ , grep("Contr_5382", x = colnames(data))] <- NULL

data <- filter(data, Reverse != "+", Potential.contaminant != "+")
data <- make_unique(data, "Gene.names", "Protein.IDs", delim = ";")
LFQ_columns <- grep("Intensity.", colnames(data))
label = gsub("Peptides.", "", colnames(data)[13:18])
experimental_design <- data.frame(
  label = label, condition = c(rep("Contr", 3), rep("AD", 3)), replicate = label
)

data <- make_se(data, LFQ_columns, experimental_design)
data
```

## QC

```{r, outlier2-qc}
plot_numbers(data)
plot_frequency(data)
plot_missval(data)
data <- filter_proteins(data, "condition", thr = 1)
plot_numbers(data)
plot_frequency(data)
plot_missval(data)

meanSdPlot(data)
data_norm <- normalize_vsn(data)
plot_normalization(data, data_norm)
data <- data_norm
rm(data_norm)
meanSdPlot(data)

plot_detect(data)
set.seed(1)
data_imp <- impute(data, fun = "QRILC")
plot_imputation(data, data_imp)
data <- data_imp
rm(data_imp)
```

## Differential Expression

```{r, outlier2-de}
data <- add_rejections(test_diff(data, type = "control", control = "Contr"), lfc = 0)
data_results <- get_results(data)
nsig <- data_results %>% filter(significant) %>% nrow()

plot_pca(data, n = round(dim(data)[1] * 0.1)) +
  scale_shape_manual(values = seq(0, length(label)))
plot_heatmap(data, type = "centered", kmeans = TRUE, k = round((nsig -1) / 2), row_font_size = 9)
plot_cor(data, pal = "RdBu", pal_rev = TRUE)
plot_volcano(data, contrast = "AD_vs_Contr", adjusted = TRUE, label_size = 4)

plot_single(data, proteins = c("MAPT", "APOE"), type = "centered") +
  xlab(element_blank())

datatable_custom(data_results)
```

# MAPT Outliers

We try removing outliers of the expected difference in MAPT.
In total, we remove `AD_5398`, `AD_5365`, `Contr_159`.

```{r, MAPT-prep}
data <- read.csv(
  file.path(
    assets_dir, "Label_Free_Quantitative_Proteomics_Analyses_Results", "proteinGroups.txt"
  ),
  sep = "\t"
)

# Reverse order of sample columns to make DE analysis more intuitive later.
col_start <- grep("AD_5333", x = colnames(data))
for (i in 1:length(col_start)) {
  section1_end <- col_start[i] - 1
  col_end <- col_start[i] + 8
  section2_start <- col_end + 1
  data <- data.frame(
    data[ , 1:section1_end], rev(data[ , col_start[i]:col_end]), data[ , section2_start:ncol(data)]
  )
}

data[ , grep("AD_5398", x = colnames(data))] <- NULL
data[ , grep("AD_5365", x = colnames(data))] <- NULL
data[ , grep("Contr_159", x = colnames(data))] <- NULL

data <- filter(data, Reverse != "+", Potential.contaminant != "+")
data <- make_unique(data, "Gene.names", "Protein.IDs", delim = ";")
LFQ_columns <- grep("Intensity.", colnames(data))
label = gsub("Peptides.", "", colnames(data)[13:18])
experimental_design <- data.frame(
  label = label, condition = c(rep("Contr", 3), rep("AD", 3)), replicate = label
)

data <- make_se(data, LFQ_columns, experimental_design)
data
```

## QC

```{r, MAPT-qc}
plot_numbers(data)
plot_frequency(data)
plot_missval(data)
data <- filter_proteins(data, "condition", thr = 1)
plot_numbers(data)
plot_frequency(data)
plot_missval(data)

meanSdPlot(data)
data_norm <- normalize_vsn(data)
plot_normalization(data, data_norm)
data <- data_norm
rm(data_norm)
meanSdPlot(data)

plot_detect(data)
set.seed(1)
data_imp <- impute(data, fun = "QRILC")
plot_imputation(data, data_imp)
data <- data_imp
rm(data_imp)
```

## Differential Expression

```{r, MAPT-de}
data <- add_rejections(test_diff(data, type = "control", control = "Contr"), lfc = 0)
data_results <- get_results(data)
nsig <- data_results %>% filter(significant) %>% nrow()

plot_pca(data, n = round(dim(data)[1] * 0.1)) +
  scale_shape_manual(values = seq(0, length(label)))
plot_heatmap(data, type = "centered", kmeans = TRUE, k = round((nsig -1) / 2), row_font_size = 9)
plot_cor(data, pal = "RdBu", pal_rev = TRUE)
plot_volcano(data, contrast = "AD_vs_Contr", adjusted = TRUE, label_size = 4)

plot_single(data, proteins = c("MAPT", "APOE"), type = "centered") +
  xlab(element_blank())

datatable_custom(data_results)
```

# References

This is the concluding section of the document. Here we write relevant results to disk, output the `sessionInfo`, and create a bibliography for works cited.

```{r, references}
sessionInfo()
```
