---
title: "Combined AD CTRL"
author:
  - name: "Emir, Pal, and Steph"
date: '`r strftime(Sys.time(), "%B %d, %Y")`'
bibliography: "../AD-CBD-exosome-proteins.bib"
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "results", "combined-AD-CTRL.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [AD-CBD-exosome-proteins](https://github.com/eturkes/AD-CBD-exosome-proteins).*
*Please email me for access.*

The data here is derived from @combined and will be referenced using the name `combined`.

```{r}
# Copyright 2019-2020 Emir Turkes, Stephanie Fowler, Guar Pallavi, UK DRI at UCL,
# Columbia University Medical Center
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# This section should be checked per document.
# --------------------------------------------
options(stringsAsFactors = FALSE)
packages <- c(
  "RColorBrewer", "dplyr", "DEP", "SummarizedExperiment", "ggplot2",
  "DT", "ComplexHeatmap", "GSEABase", "GSVA", "limma"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path(getwd(), "utils.R"))

data_name <- "combined"
color <- colorRampPalette(rev(brewer.pal(9, "RdBu")))(100)
`%notin%` <- Negate(`%in%`)
# --------------------------------------------

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------
assets_dir <- file.path(getwd(), "..", "assets") # Backed up data.

cache_dir <- file.path(getwd(), "..", "tmp", "cache")
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

knitr::opts_chunk$set(fig.width = 10, fig.height = 7)
# ----------------------------------------------------------
```

# All Data

## Prep

```{r}
data <- read.csv(file.path(assets_dir, data_name, "data_comb_steph.csv"), row.names = "Gene.names_")
design <- read.csv(file.path(assets_dir, data_name, "comb_steph_newexp_design.csv"))

data <- make_unique(data, "Gene.names", "Protein.IDs.x", ";")
LFQ_columns <- grep("Intensity.", colnames(data))

data <- make_se(data, LFQ_columns, design)
data
```

**Unaltered data, already log2 transformed.**

```{r}
datatable_download(assay(data))
```

**After removal of log transformation.**

```{r}
datatable_download(2 ^ assay(data))
```

## QC

### Histogram

```{r}
hist(assay(data), n = 100)
```

### Missing Value Removal

```{r}
discard <- which(is.na(rowMeans(assay(data)[ , 1:ceiling(ncol(data) / 2)], TRUE)))
data <- data[-discard, ]
discard <- which(is.na(rowMeans(assay(data)[ , (ceiling(ncol(data) / 2) + 1):ncol(data)], TRUE)))
data <- data[-discard, ]
complete_cases <- filter_proteins(data, "complete")

plot_numbers(data)
plot_frequency(data)
plot_detect(data)
```

```{r, fig.height = 100}
AD_order <- order(rowMeans(assay(data)[ , 1:ceiling(ncol(data) / 2)], TRUE))
data <- data[AD_order, ]
AD_MNAR <- rownames(data)[1:which(rownames(data) == "ATRN")] # TMX4
plot_missval_custom(data)

Control_order <- order(rowMeans(assay(data)[ , (ceiling(ncol(data) / 2) + 1):ncol(data)], TRUE))
data <- data[Control_order, ]
Control_MNAR <- rownames(data)[1:which(rownames(data) == "PMP2")] # MGST1
plot_missval_custom(data)

AD_data <- data[ , 1:ceiling(ncol(data) / 2)]
AD_data <- AD_data[rownames(AD_data) %notin% AD_MNAR, ]
missing_data <- assay(AD_data) %>% data.frame(.)
missing_data <- ifelse(is.na(missing_data), 0, 1)
discard <- which(rowSums(missing_data, TRUE) < 5)
AD_data <- AD_data[-discard, ]
print(paste0("Removed ", length(discard), " proteins from AD group."))

Control_data <- data[ , (ceiling(ncol(data) / 2) + 1):ncol(data)]
Control_data <- Control_data[rownames(Control_data) %notin% Control_MNAR, ]
missing_data <- assay(Control_data) %>% data.frame(.)
missing_data <- ifelse(is.na(missing_data), 0, 1)
discard <- which(rowSums(missing_data, TRUE) < 5)
Control_data <- Control_data[-discard, ]
print(paste0("Removed ", length(discard), " proteins from Control group."))

AD_MAR <- rownames(AD_data)[
  rownames(AD_data) %in% Control_MNAR | rownames(AD_data) %in% rownames(Control_data)
]
Control_MAR <- rownames(Control_data)[
  rownames(Control_data) %in% AD_MNAR | rownames(Control_data) %in% rownames(AD_data)
]
rm(AD_data, Control_data)
MAR <- unique(append(AD_MAR, Control_MAR))

AD_MNAR_passQC <- AD_MNAR[AD_MNAR %in% Control_MAR | AD_MNAR %in% Control_MNAR]
Control_MNAR_passQC <- Control_MNAR[Control_MNAR %in% AD_MAR | Control_MNAR %in% AD_MNAR]
MNAR <- unique(append(AD_MNAR_passQC, Control_MNAR_passQC))

keep <- unique(append(MAR, MNAR))
data <- data[rownames(data) %in% keep, ]

data <- data[sort(rownames(data)), ]
AD_order <- order(rowMeans(assay(data)[ , 1:ceiling(ncol(data) / 2)], TRUE))
Control_order <- order(rowMeans(assay(data)[ , (ceiling(ncol(data) / 2) + 1):ncol(data)], TRUE))
missing_data <- assay(data) %>% data.frame(.)
missing_data <- missing_data[apply(missing_data, 1, function(x) any(is.na(x))), ]
```

```{r}
plot_numbers(data)
plot_frequency(data)
plot_detect(data)
```

**Data before normalization but after removal of missing values.**

```{r}
datatable_download(assay(data))
```

**Complete cases before normalization but after removal of missing values.**

```{r}
datatable_download(assay(complete_cases))
```

### Normalization

```{r}
VSN_Normalized <- normalize_vsn(data)
Log_Transform_Only <- data

VSN_Normalized_Non_Missing_Data <- normalize_vsn(complete_cases)
Log_Transform_Only_Non_Missing_Data <- complete_cases

print("Log_Transform_Only")
meanSdPlot(Log_Transform_Only)
print("Log_Transform_Only_Non_Missing_Data")
meanSdPlot(Log_Transform_Only_Non_Missing_Data)
print("VSN_Normalized")
meanSdPlot(VSN_Normalized)
print("VSN_Normalized_Non_Missing_Data")
meanSdPlot(VSN_Normalized_Non_Missing_Data)

plot_normalization(Log_Transform_Only, VSN_Normalized)
plot_normalization(Log_Transform_Only_Non_Missing_Data, VSN_Normalized_Non_Missing_Data)

data <- VSN_Normalized
complete_cases <- VSN_Normalized_Non_Missing_Data
rm(
  VSN_Normalized, VSN_Normalized_Non_Missing_Data,
  Log_Transform_Only, Log_Transform_Only_Non_Missing_Data
)
```

**Data after normalization and log2 transformation.**

```{r}
datatable_download(assay(data))
```

### Missing Value Imputation

```{r}
unimputed_data <- data
Unimputed_Data <- unimputed_data
Non_Missing_data <- complete_cases
plot_imputation(Unimputed_Data, Non_Missing_data)

impute1 <- data[ , 1:ceiling(ncol(data) / 2)]
impute_vector <- rownames(impute1) %in% AD_MNAR_passQC
set.seed(1)
impute1 <- impute(impute1, "mixed", randna = !impute_vector, mar = "knn", mnar = "MinDet")

impute2 <- data[ , (ceiling(ncol(data) / 2) + 1):ncol(data)]
impute_vector <- rownames(impute2) %in% Control_MNAR_passQC
set.seed(1)
impute2 <- impute(impute2, "mixed", randna = !impute_vector, mar = "knn", mnar = "MinDet")

Imputed_Data <- data
assay(Imputed_Data) <- cbind(assay(impute1), assay(impute2))

plot_imputation(Imputed_Data)

data <- Imputed_Data
rm(impute1, impute2, Non_Missing_data)
```

```{r, fig.height = 100}
data <- data[sort(rownames(data)), ]
data <- data[AD_order, ]
heatmap_data <- data[rownames(data) %in% rownames(missing_data), ]
Heatmap(
  assay(heatmap_data),
  colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(round(max(assay(heatmap_data)))),
  column_names_side = "top",
  show_column_names = TRUE,
  cluster_columns = FALSE,
  cluster_rows = FALSE,
  row_names_side = "left",
  heatmap_legend_param = list(
    title = "log2 Intensity", title_position = "leftcenter-rot"
  ),
  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 16)
)

data <- data[sort(rownames(data)), ]
data <- data[Control_order, ]
heatmap_data <- data[rownames(data) %in% rownames(missing_data), ]
Heatmap(
  assay(heatmap_data),
  colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(round(max(assay(heatmap_data)))),
  column_names_side = "top",
  show_column_names = TRUE,
  cluster_columns = FALSE,
  cluster_rows = FALSE,
  row_names_side = "left",
  heatmap_legend_param = list(
    title = "log2 Intensity", title_position = "leftcenter-rot"
  ),
  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 16)
)
```

```{r}
plot_normalization(Unimputed_Data, Imputed_Data)
print("Unimputed Data")
meanSdPlot(Unimputed_Data)
print("Imputed Data")
meanSdPlot(Imputed_Data)

rm(Imputed_Data, Unimputed_Data)
```

**Data after missing value imputation.**
**This is final form used for differential expression analysis**

```{r}
datatable_download(assay(data))
```

### PCA

```{r}
plot_pca(data, n = dim(data)[1], indicate = c("condition", "label"), label = TRUE) +
  scale_shape_manual(values = seq(0, ncol(data))) +
  ggtitle("PCA Plot")
```

## Differential Expression

The results below are filtered for significance ($adjP < 0.05$).

```{r}
data_results <- add_rejections(test_diff(data, "control", "Control"), lfc = 0)
results <- get_results(data_results)
sig <- results[results$significant, ]
print(paste0(nrow(sig), " significant proteins."))

plot_heatmap_custom(data_results, "centered", row_font_size = 9)
plot_heatmap(data_results, "centered", row_font_size = 9)
plot_cor(data_results, pal = "RdBu", pal_rev = TRUE)
plot_volcano(data_results, "AD_vs_Control", 4, adjusted = TRUE)
plot_single(data_results, sig$name, "centered") + xlab(element_blank())

datatable_download(results)
```

```{r, fig.height = 100}
Heatmap(
  assay(data),
  colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(round(max(assay(data)))),
  column_names_side = "top",
  show_column_names = TRUE,
  cluster_columns = FALSE,
  row_names_side = "left",
  heatmap_legend_param = list(
    title = "log2 Intensity", title_position = "leftcenter-rot"
  ),
  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 16)
)

no_mult_testing <- results[results$AD_vs_Control_p.val < 0.05, ]
no_mult_testing <- data_results[rownames(data_results) %in% no_mult_testing$name, ]
Heatmap(
  assay(no_mult_testing),
  colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(round(max(assay(no_mult_testing)))),
  column_names_side = "top",
  show_column_names = TRUE,
  cluster_columns = FALSE,
  row_names_side = "left",
  heatmap_legend_param = list(
    title = "log2 Intensity", title_position = "leftcenter-rot"
  ),
  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 16)
)

missing_no_mult_testing <- no_mult_testing[rownames(no_mult_testing) %in% rownames(missing_data), ]
heatmap <- Heatmap(
  assay(missing_no_mult_testing),
  colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(round(max(assay(missing_no_mult_testing)))),
  column_names_side = "top",
  show_column_names = TRUE,
  cluster_columns = FALSE,
  row_names_side = "left",
  heatmap_legend_param = list(
    title = "log2 Intensity", title_position = "leftcenter-rot"
  ),
  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 16)
)
draw(heatmap)

unimputed_missing_no_mult_testing <- unimputed_data[
  rownames(unimputed_data) %in% rownames(missing_no_mult_testing),
]
Heatmap(
  assay(unimputed_missing_no_mult_testing),
  colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(round(max(assay(missing_no_mult_testing)))),
  column_names_side = "top",
  show_column_names = TRUE,
  cluster_columns = FALSE,
  row_order = row_order(heatmap),
  row_names_side = "left",
  heatmap_legend_param = list(
    title = "log2 Intensity", title_position = "leftcenter-rot"
  ),
  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 16)
)
rm(unimputed_missing_no_mult_testing, missing_no_mult_testing)
```

## Non-Missing Data

### PCA

```{r}
plot_pca(
  complete_cases, n = dim(complete_cases)[1], indicate = c("condition", "label"), label = TRUE
  ) + scale_shape_manual(values = seq(0, ncol(data))) +
  ggtitle("PCA Plot")
```

### Differential Expression

```{r}
complete_results <- add_rejections(test_diff(complete_cases, "control", "Control"), lfc = 0)
results <- get_results(complete_results)
sig <- results[results$significant, ]
print(paste0(nrow(sig), " significant proteins."))

plot_heatmap_custom(complete_results, "centered", row_font_size = 9)
plot_heatmap(complete_results, "centered", row_font_size = 9)
plot_cor(complete_results, pal = "RdBu", pal_rev = TRUE)
plot_volcano(complete_results, "AD_vs_Control", label_size = 4, adjusted = TRUE)
plot_single(complete_results, sig$name, "centered") + xlab(element_blank())

datatable_download(results)
```

```{r, fig.height = 100}
Heatmap(
  assay(complete_results),
  colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(round(max(assay(complete_results)))),
  column_names_side = "top",
  show_column_names = TRUE,
  cluster_columns = FALSE,
  row_names_side = "left",
  heatmap_legend_param = list(
    title = "log2 Intensity", title_position = "leftcenter-rot"
  ),
  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 16)
)

no_mult_testing <- results[results$AD_vs_Control_p.val < 0.05, ]
no_mult_testing <- complete_results[rownames(complete_results) %in% no_mult_testing$name, ]
Heatmap(
  assay(no_mult_testing),
  colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(round(max(assay(no_mult_testing)))),
  column_names_side = "top",
  show_column_names = TRUE,
  cluster_columns = FALSE,
  row_names_side = "left",
  heatmap_legend_param = list(
    title = "log2 Intensity", title_position = "leftcenter-rot"
  ),
  row_names_gp = gpar(fontsize = 6),
  column_names_gp = gpar(fontsize = 16)
)
rm(no_mult_testing)
```

## Gene Set Enrichment

### Johnson et al. 2020 Modules

```{r}
gene_sets <- getGmt(file.path(assets_dir, "gene-sets", "proteomics-modules.gmt"))
gene_sets

data_filt <- data[rownames(data) %in% unique(unlist(geneIds(gene_sets))), ]
print(paste0("Unique proteins in post-QC protein matrix: ", dim(data)[1]))
print(paste0("Unique proteins in proteomics modules: ", length(unique(unlist(geneIds(gene_sets))))))
print(
  paste0("Unique proteins in post-QC protein matrix within proteomics modules: ", dim(data_filt)[1])
)

gsva <- gsva(assay(data_filt), gene_sets, verbose = FALSE)
datatable_download(gsva)

Heatmap(
  gsva,
  color,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(gsva)),
  column_names_side = "top",
  heatmap_legend_param = list(
    title = "GSVA enrichment", title_position = "leftcenter-rot"
  ),
)

limma_design <- model.matrix(~ 0 + design$condition, as.data.frame(gsva))
colnames(limma_design) <- unique(design$condition)
fit <- lmFit(gsva, limma_design)
contrast_mat <- makeContrasts(AD-Control, levels = limma_design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit)
results <- topTable(cont_fit, 1, Inf)

datatable_download_exp(results)
```

### Panglao Cell Types

```{r}
gene_sets <- getGmt(file.path(assets_dir, "gene-sets", "panglao-brain-ctypes.gmt"))
gene_sets

data_filt <- data[rownames(data) %in% unique(unlist(geneIds(gene_sets))), ]
print(paste0("Unique proteins in post-QC protein matrix: ", dim(data)[1]))
print(paste0("Unique proteins in proteomics modules: ", length(unique(unlist(geneIds(gene_sets))))))
print(
  paste0("Unique proteins in post-QC protein matrix within proteomics modules: ", dim(data_filt)[1])
)

gsva <- gsva(assay(data_filt), gene_sets, verbose = FALSE)
datatable_download(gsva)

Heatmap(
  gsva,
  color,
  cluster_columns = FALSE,
  row_names_max_width = max_text_width(rownames(gsva)),
  column_names_side = "top",
  heatmap_legend_param = list(
    title = "GSVA enrichment", title_position = "leftcenter-rot"
  ),
)

limma_design <- model.matrix(~ 0 + design$condition, as.data.frame(gsva))
colnames(limma_design) <- unique(design$condition)
fit <- lmFit(gsva, limma_design)
contrast_mat <- makeContrasts(AD-Control, levels = limma_design)
cont_fit <- eBayes(contrasts.fit(fit, contrast_mat))
tests <- decideTests(cont_fit)
results <- topTable(cont_fit, 1, Inf)

datatable_download_exp(results)
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
