---
title: "Marker Gene Analysis"
author:
  - name: "Emir Turkes and Stephanie Fowler"
date: '`r strftime(Sys.time(), "%B %d, %Y")`'
bibliography: "../AD-exosome-characterisation.bib"
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(
    "..", "results", "marker_gene_analysis.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [AD-exosome-characterisation](https://github.com/eturkes/AD-exosome-characterisation).*
*Please email for access.*

The data here will be referenced using the name `markers`.
This analysis is for the purpose of deriving marker genes from human cortical single-nucleus RNAseq data from the Allen Brain Atlas.

```{r}
# Some standard boilerplate.
# --------------------------
#    This file is part of AD-exosome-characterisation.
#    Copyright (C) 2022-2023  Emir Turkes, Stephanie Fowler, UK DRI at UCL, Columbia
#    University Medical Center
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# This section should be checked per document.
# --------------------------------------------
packages <- c("conflicted", "Seurat", "Matrix", "parallelly", "DT")
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path("utils.R"))
# --------------------------------------------

# Everything else in this chunk remains generally unchanged.
# ----------------------------------------------------------
assets_dir <- file.path("..", "assets") # Backed up data.

cache_dir <- file.path("..", "tmp", "cache", "markers")
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dev = "svglite")
# ----------------------------------------------------------
# --------------------------
```

# Prep

```{r}
rds <- file.path(cache_dir, "data.rds")
if (file.exists(rds)) {
  data <- readRDS(rds)
} else {

  data <- read.csv(file.path(assets_dir, "allen-human-cortical-tx", "exon.csv"))
  meta <- read.csv(file.path(assets_dir, "allen-human-cortical-tx", "metadata.csv"))

  rownames(meta) <- meta$exp_component_name
  meta$exp_component_name <- NULL
  rownames(meta) <- gsub("-", ".", rownames(meta))
  rownames(data) <- data$X
  data <- data[ , -1]
  data <- data[ , colnames(data) %in% rownames(meta)]
  data <- as.matrix(data)
  data <- Matrix(data, sparse = TRUE)

  data <- CreateSeuratObject(data, meta.data = meta)
  rm(meta)

  remove <- which(data$subclass_label == "")
  data <- data[ , -remove]

  data <- SCTransform(data, conserve.memory = TRUE, vst.flavor = "v2")

  saveRDS(data, rds)
}
```

# Analysis

```{r}
tsv <- file.path(assets_dir, "gene-sets", "markers.tsv")
if (file.exists(rds)) {
  markers <- read.delim(tsv)
} else {

  neuronal <- data[ , which(data$class_label != "Non-neuronal")]
  non_neuronal <- data[ , which(data$class_label == "Non-neuronal")]

  data <- data[ , c(colnames(neuronal), colnames(non_neuronal))]
  data$class_subclass <- c(neuronal$class_label, non_neuronal$subclass_label)
  rm(neuronal, non_neuronal)
  data@active.ident <- factor(data$class_subclass)

  # Set up parallel processing.
  # Not working reliably, so commented out.
  # ---------------------------------------
  # gc()
  # mem <- as.numeric(unlist(strsplit(system("free -b", TRUE)[2], " "))[8])
  # mem <- mem - 10 * 1024 ^ 3
  # mem <- mem / availableCores()
  # options(future.globals.maxSize = mem)
  # plan("multisession")
  # ---------------------------------------

  markers <- FindAllMarkers(data, only.pos = TRUE)

  write.table(markers, file.path(assets_dir, "gene-sets", "markers.tsv"), sep = "\t")
}

datatable_download_exp(markers)
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
